jumpscale/sals/nettools/__init__.py

```diff
@@ -375,98 +557,262 @@ def get_host_by_name(dnsHostname: str):
-    @type username: string
-    @param passwd: password for the url if it requires authentication
-    @type passwd: string
+
+    Args:
+        url (str): URL to download from
+        localpath (str): filename or directory to download the url to. pass None to return the data. Defaults to ''.
+        username (str, optional): username for the url if it requires authentication. Defaults to None.
+        passwd (str, optional): password for the url if it requires authentication. Defaults to None.
+        overwrite (bool, optional): if the file exists, it will be truncated. Defaults to True.
+        append_to_home (bool, optional): if set to true, any relative path specified in localpath arg
+            will be appended to the user home directory (that guaranteed to have a write permission). if set
+            to False any relative localpath set by user will append to the current working directory. if user
+            specified a absolute localpath (start with /) append_to_home will have no effect. Defaults to False.
+        name_from_url (bool, optional): if set to true, localpath will treated as a dir, and will try to get
+            the file name from the url or fallback to auto generated name. Defaults to True.
+
+    Raises:
+        PermissionError: [description]
+        FileNotFoundError: [description]
+        FileExistsError: [description]
+        ValueError: [description]
+        URLError: [description]
+
+    Returns:
+        namedtuple: namedtuple('DownloadResult', ['localpath', 'content', content_length])
+            - localpath (pathlib.Path)
+            - content (bytes): only if localpath is None else it will be None always
+            - content_length: the content length as returned from the response headers
+
+    Todo:
+        - better performance with Multi-Threaded
+        - support resume
+
+    Examples:
+        # use default values for args will download the url to cwd and get the name from the url,
+        # if the file already exists, it will overwritten.
+        >>> nettools.download('https://www.7-zip.org/a/7z1900-extra.7z')
+        DownloadResult(localpath=PosixPath('/home/sameh/projects/js-ng/7z1900-extra.7z'), content=None, content_length='929117')
+
     """
-    if not url:
-        raise Value("URL can not be None or empty string")
-    if not localpath:
-        raise Value("Local path to download the url to can not be None or empty string")
-    filename = ""
-    if localpath == "-":
-        filename = "-"
-    if jumpscale.sals.fs.exists(localpath) and jumpscale.sals.fs.is_dir(localpath):
-        filename = jumpscale.sals.fs.join_paths(localpath, jumpscale.sals.fs.basename(url))
-    else:
-        if jumpscale.sals.fs.is_dir(jumpscale.sals.fs.dirname(localpath)):
-            filename = localpath
-        else:
-            raise Value("Local path is an invalid path")
+    DownloadResult = namedtuple("DownloadResult", ["localpath", "content", "content_length"])
```

Comment:
can be moved outside in the global scope

Reply:
> can be moved outside in the global scope
I don't agree here, this doesn't need to be available anywhere outside the function.
this way keeps the function easier to maintain, and the global scope as clean as possible.